databus/src/main/java/com/bazaarvoice/emodb/databus/core/DefaultDatabus.java

```diff
@@ -740,7 +780,111 @@ private void checkSubscriptionOwner(String ownerId, OwnedSubscription subscripti
+            }
+        });
+    }
+
+    private void doDrainQueue(String subscription, int itemsToFetch) {
+        boolean nonRedundantItemFound = false;
+        Stopwatch stopwatch = Stopwatch.createStarted(_ticker);
+
+        ConsolidatingEventSink sink = new ConsolidatingEventSink(itemsToFetch);
+        boolean more = _eventStore.peek(subscription, sink);
+
+        Map<Coordinate, EventList> rawEvents = sink.getEvents();
+
+        if (rawEvents.isEmpty()) {
+            return;  // queue is empty.
+        }
+
+        List<String> eventIdsToDiscard = Lists.newArrayList();
+
+        // Query the events from the data store in batch to reduce latency.
+        DataProvider.AnnotatedGet annotatedGet = _dataProvider.prepareGetAnnotated(ReadConsistency.STRONG);
+        for (Map.Entry<Coordinate, EventList> entry : rawEvents.entrySet()) {
+            Coordinate coord = entry.getKey();
+
+            // Query the table/key pair.
+            try {
+                annotatedGet.add(coord.getTable(), coord.getId());
+            } catch (UnknownTableException e) {
+                // It's likely the table was dropped since the event was queued.  Discard the events.
+                EventList list = entry.getValue();
+                for (Pair<String, UUID> pair : list.getEventAndChangeIds()) {
+                    eventIdsToDiscard.add(pair.first());
+                }
+            }
+        }
+        Iterator<DataProvider.AnnotatedContent> readResultIter = annotatedGet.execute();
+
+        // Loop through the results of the data store query.
+        while (readResultIter.hasNext()) {
+            DataProvider.AnnotatedContent readResult = readResultIter.next();
+
+            // Get the JSON System of Record entity for this piece of content
+            Map<String, Object> content = readResult.getContent();
+
+            // Find the original event IDs that correspond to this piece of content
+            Coordinate coord = Coordinate.fromJson(content);
+            EventList eventList = rawEvents.get(coord);
+
+            // Loop over the Databus events for this piece of content.  Usually there's just one, but not always...
+            for (Pair<String, UUID> eventData : eventList.getEventAndChangeIds()) {
+                String eventId = eventData.first();
+                UUID changeId = eventData.second();
+
+                // Is the change redundant?
+                if (readResult.isChangeDeltaRedundant(changeId)) {
+                    eventIdsToDiscard.add(eventId);
+                    continue;
+                }
+
+                nonRedundantItemFound = true;
+            }
+        }
+
+        // delete the events we never again want to see.
+        if (!eventIdsToDiscard.isEmpty()) {
+            _eventStore.delete(subscription, eventIdsToDiscard, true);
+        }
+
+        long totalSubscriptionDrainTime = _drainedSubscriptionsMap.get(subscription).longValue() + stopwatch.elapsed(TimeUnit.MILLISECONDS);
```

Comment:
Do we want 1 minute total time draining the queue or 1 minute wall clock time since we first started draining?